src/mbgl/util/mapbox.cpp

```diff
@@ -39,6 +39,20 @@ std::string normalizeGlyphsURL(const std::string& url, const std::string& access
     return normalizeURL(url, accessToken);
 }
 
+std::string normalizeTileURL(const std::string& url) {
+    if (url.compare(url.length() - 4, 4, ".png") == 0)
```

Comment:
It's also important that this only insert `{ratio}` on Mapbox sources, since we can't guarantee that other sources support `@2x`. That's why [gl-js check the source URL](https://github.com/mapbox/mapbox-gl-js/blob/8ee8e879d6863821278dbdd99ab427458b0f04ee/js/util/mapbox.js#L62) against `mapbox://`.